include/swift/Runtime/Concurrency.h

```diff
@@ -151,39 +167,116 @@ SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swiftasync)
 TaskFutureWaitThrowingSignature::FunctionType
 swift_task_future_wait_throwing;
 
+using TaskGroupFutureWaitThrowingSignature =
+  AsyncSignature<void(AsyncTask *, TaskGroup *, Metadata *), /*throws*/ true>;
+
 /// Wait for a readyQueue of a Channel to become non empty.
 ///
 /// This can be called from any thread. Its Swift signature is
 ///
 /// \code
-/// func swift_task_group_wait_next(on groupTask: Builtin.NativeObject) async
-///     -> (hadErrorResult: Bool, storage: UnsafeRawPointer?)
+/// func swift_task_group_wait_next_throwing(
+///     waitingTask: Builtin.NativeObject, // current task
+///     group: Builtin.NativeObject,
+/// ) async -> T
 /// \endcode
 SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swiftasync)
-TaskFutureWaitSignature::FunctionType
-swift_task_group_wait_next;
+TaskGroupFutureWaitThrowingSignature::FunctionType
+swift_task_group_wait_next_throwing;
+
+/// Create a new `TaskGroup` using the task's allocator.
+/// The caller is responsible for retaining and managing the group's lifecycle.
+///
+/// Its Swift signature is
+///
+/// \code
+/// func swift_task_group_create(
+///     _ task: Builtin.NativeObject
+/// ) -> Builtin.NativeObject
+/// \endcode
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+swift::TaskGroup* swift_task_group_create(AsyncTask *task);
+
+/// Attach a child task to the parent task's task group record.
+///
+/// Its Swift signature is
+///
+/// \code
+/// func swift_task_group_attachChild(
+///     group: Builtin.NativeObject,
+///     parent: Builtin.NativeObject,
+///     child: Builtin.NativeObject
+/// )
+/// \endcode
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+void swift_task_group_attachChild(TaskGroup *group,
+                                  AsyncTask *parent, AsyncTask *child);
 
+/// Its Swift signature is
+///
+/// \code
+/// func swift_task_group_destroy(
+///     _ task: Builtin.NativeObject,
+///     _ group: Builtin.NativeObject
+/// )
+/// \endcode
+SWIFT_EXPORT_FROM(swift_Concurrency) SWIFT_CC(swift)
+void swift_task_group_destroy(AsyncTask *task, TaskGroup *group);
```

Comment:
If `TaskGroup` is your C++ class defined in TaskGroup.h, it does not seem to be a `NativeObject`.

Reply:
Yeah, it's an instance of that C++ class.

It is used like this:

```
    let _group = _taskGroupCreate(task: task)
    _taskGroupDestroy(task: task, group: _group)

/// ---

@_silgen_name("swift_task_group_create")
func _taskGroupCreate(
  task: Builtin.NativeObject
) -> Builtin.NativeObject

@_silgen_name("swift_task_group_destroy")
func _taskGroupDestroy(
  task: Builtin.NativeObject,
  group: __owned Builtin.NativeObject
)
```

What should it be instead?

Reply:
Solved: OpaqueValue* is correct, will fix